verilog/tools/kythe/kythe_facts_extractor.cc

```diff
@@ -0,0 +1,230 @@
+      file_path_,
+      CreateScopeRelativeSignature(CreateModuleSignature(module_name.Value())));
+  const VName module_name_anchor = PrintAnchorVName(module_name, file_path_);
+
+  *stream_ << Fact(module_vname, kFactNodeKind, kNodeRecord);
+  *stream_ << Fact(module_vname, kFactSubkind, kSubkindModule);
+  *stream_ << Fact(module_vname, kFactComplete, kCompleteDefinition);
+  *stream_ << Edge(module_name_anchor, kEdgeDefinesBinding, module_vname);
+
+  if (anchors.size() > 1) {
+    const VName module_end_label_anchor =
+        PrintAnchorVName(module_end_label, file_path_);
+    *stream_ << Edge(module_end_label_anchor, kEdgeRef, module_vname);
+  }
+
+  return module_vname;
+}
+
+VName KytheFactsExtractor::ExtractModuleInstanceFact(
+    const IndexingFactNode& node) {
+  const auto& anchors = node.Value().Anchors();
+  const Anchor& module_type = anchors[0];
+  const Anchor& instance_name = anchors[1];
+
+  const VName module_instance_vname(
+      file_path_, CreateScopeRelativeSignature(
+                      CreateVariableSignature(instance_name.Value())));
+  const VName module_instance_anchor =
+      PrintAnchorVName(instance_name, file_path_);
+
+  const VName module_type_vname =
+      *ABSL_DIE_IF_NULL(scope_context_.SearchForDefinition(
+          CreateModuleSignature(module_type.Value())));
+
+  const VName module_type_anchor = PrintAnchorVName(module_type, file_path_);
+
+  *stream_ << Fact(module_instance_vname, kFactNodeKind, kNodeVariable);
+  *stream_ << Fact(module_instance_vname, kFactComplete, kCompleteDefinition);
+
+  *stream_ << Edge(module_type_anchor, kEdgeRef, module_type_vname);
+  *stream_ << Edge(module_instance_vname, kEdgeTyped, module_type_vname);
+  *stream_ << Edge(module_instance_anchor, kEdgeDefinesBinding,
+                   module_instance_vname);
+
+  for (const auto& anchor :
+       verible::make_range(anchors.begin() + 2, anchors.end())) {
+    const VName port_vname_reference(
+        file_path_,
+        CreateScopeRelativeSignature(CreateVariableSignature(anchor.Value())));
+    const VName port_vname_definition(
+        file_path_,
+        CreateScopeRelativeSignature(CreateVariableSignature(anchor.Value())));
+    const VName port_vname_anchor = PrintAnchorVName(anchor, file_path_);
+
+    *stream_ << Edge(port_vname_anchor, kEdgeRef, port_vname_definition);
+  }
+
+  return module_instance_vname;
+}
+
+VName KytheFactsExtractor::ExtractVariableDefinition(
+    const IndexingFactNode& node) {
+  const auto& anchor = node.Value().Anchors()[0];
+  const VName variable_vname(
+      file_path_,
+      CreateScopeRelativeSignature(CreateVariableSignature(anchor.Value())));
+  const VName variable_vname_anchor = PrintAnchorVName(anchor, file_path_);
+
+  *stream_ << Fact(variable_vname, kFactNodeKind, kNodeVariable);
+  *stream_ << Fact(variable_vname, kFactComplete, kCompleteDefinition);
+
+  *stream_ << Edge(variable_vname_anchor, kEdgeDefinesBinding, variable_vname);
+
```

Comment:
@fangism can you give some example?
I didn't get what ur thinking of but I would like to know. Just because I like designs likes this :)

Reply:
One example of this are the `Apply*` functions in https://github.com/google/verible/blob/master/common/util/vector_tree.h.

If I want to collect pointers to elements in post-order, I could:

```c++
vector<const Foo*> f;
// vector_tree is a VectorTree<Foo>
vector_tree.ApplyPostOrder([&f](const Foo& value) { f.push_back(&value); });
```

To print in post-order:

```c++
vector_tree.ApplyPostOrder([&f](const Foo& value) { std::cout << value; });
```

If you expose an interface to your extractor to operate on any Fact/Edge, it might look like:

```c++
class Extractor {
  void Walk(const std::function<void(const Fact&)>& fact_listener,
            const std::function<void(const Edge&)>& edge_listener) const;
};
```